File: iosApp/TTCoach/Pose/ORTSessionFactory.swift
// ORTSessionFactory.swift
//
// Small reusable helper that builds configured ONNX Runtime sessions. Used by both
// the in-app pose backend (iOS) and the RTMPoseExport macOS CLI spike.
//
// The ONNX Runtime SPM package (microsoft/onnxruntime-swift-package-manager, product
// "onnxruntime") exposes the Objective-C API under the Swift module `OnnxRuntimeBindings`.
//
// We share a single ORTEnv across sessions and attempt to append the CoreML Execution
// Provider; if CoreML EP is unavailable or fails to append, we fall back to plain CPU
// (ORT's default) instead of failing. The chosen path is logged.

import Foundation
import OnnxRuntimeBindings

/// Errors surfaced when building an ORT session. Swift `throws` is fine here — this is
/// platform (Apple) code, not KMP shared logic.
enum ORTSessionFactoryError: Error, CustomStringConvertible {
    case modelFileMissing(path: String)
    case sessionCreationFailed(model: String, underlying: Error)

    var description: String {
        switch self {
        case .modelFileMissing(let path):
            return "ONNX model file not found at: \(path)"
        case .sessionCreationFailed(let model, let underlying):
            return "Failed to create ORTSession for \(model): \(underlying)"
        }
    }
}

/// Builds configured `ORTSession` instances backed by a single shared `ORTEnv`.
enum ORTSessionFactory {

    /// One process-wide ORT environment (log level: warning).
    private static let sharedEnv: ORTEnv = {
        // ORTEnv init can throw; if it does at process start, the spike is fundamentally
        // broken, so a fatalError with a clear message is acceptable here.
        do {
            return try ORTEnv(loggingLevel: ORTLoggingLevel.warning)
        } catch {
            fatalError("Failed to initialize ORTEnv: \(error)")
        }
    }()

    /// Builds session options, appending the CoreML Execution Provider when available.
    /// Falls back to plain CPU (the ORT default) if CoreML EP cannot be appended.
    private static func makeSessionOptions(modelLabel: String) throws -> ORTSessionOptions {
        let options = try ORTSessionOptions()
        do {
            let coreMLOptions = ORTCoreMLExecutionProviderOptions()
            try options.appendCoreMLExecutionProvider(with: coreMLOptions)
            print("[ORTSessionFactory] \(modelLabel): CoreML Execution Provider appended.")
        } catch {
            // CoreML EP not available on this platform/build, or append failed — fall
            // back to CPU. This is not fatal; ORT runs on CPU by default.
            print("[ORTSessionFactory] \(modelLabel): CoreML EP unavailable (\(error)); falling back to CPU.")
        }
        return options
    }

    /// Creates a session for a model at an absolute file path (used by the macOS CLI).
    static func makeSession(modelPath: String, modelLabel: String? = nil) throws -> ORTSession {
        let label = modelLabel ?? (modelPath as NSString).lastPathComponent
        guard FileManager.default.fileExists(atPath: modelPath) else {
            throw ORTSessionFactoryError.modelFileMissing(path: modelPath)
        }
        let options = try makeSessionOptions(modelLabel: label)
        do {
            return try ORTSession(env: sharedEnv, modelPath: modelPath, sessionOptions: options)
        } catch {
            throw ORTSessionFactoryError.sessionCreationFailed(model: label, underlying: error)
        }
    }

    /// Creates a session for a model file URL.
    static func makeSession(modelURL: URL, modelLabel: String? = nil) throws -> ORTSession {
        try makeSession(modelPath: modelURL.path, modelLabel: modelLabel)
    }

    /// Creates a session from an app-bundle resource (used by the iOS app backend).
    static func makeSession(
        bundleResource name: String,
        extension ext: String = "onnx",
        in bundle: Bundle = .main
    ) throws -> ORTSession {
        guard let url = bundle.url(forResource: name, withExtension: ext) else {
            throw ORTSessionFactoryError.modelFileMissing(path: "\(name).\(ext) (in bundle \(bundle.bundlePath))")
        }
        return try makeSession(modelURL: url, modelLabel: name)
    }
}

File: app/src/main/java/com/ttcoachai/pose/OrtSessionFactory.kt
package com.ttcoachai.pose

import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtException
import ai.onnxruntime.OrtSession
import android.content.res.AssetManager
import java.io.File
import java.io.IOException

/**
 * Small reusable helper that builds configured ONNX Runtime sessions. 1:1 Kotlin port of
 * `iosApp/TTCoach/Pose/ORTSessionFactory.swift`.
 *
 * We share a single [OrtEnvironment] across sessions. Unlike iOS (which appends a CoreML EP and
 * falls back to CPU on failure), Android skips the accelerated-EP story here: YOLOX MUST run
 * CPU-only regardless (baked-in NMS breaks accelerated EPs — same reason iOS forces CPU for it),
 * and callers that want NNAPI for RTMPose wire it explicitly via [SessionConfig] rather than this
 * factory guessing at fallback behavior.
 */
object OrtSessionFactory {

    /** One process-wide ORT environment. */
    private val sharedEnv: OrtEnvironment by lazy { OrtEnvironment.getEnvironment() }

    /** Errors surfaced when building an ORT session. */
    sealed class OrtSessionFactoryError(message: String) : Exception(message) {
        class ModelFileMissing(path: String) :
            OrtSessionFactoryError("ONNX model file not found at: $path")

        class SessionCreationFailed(model: String, underlying: Throwable) :
            OrtSessionFactoryError("Failed to create OrtSession for $model: ${underlying.message}")
    }

    /**
     * Creates a session for a model at an absolute file path. Pass `cpuOnly = true` for a
     * CPU-only session (required for the YOLOX detector).
     */
    @Throws(OrtSessionFactoryError::class)
    fun makeSession(modelPath: String, modelLabel: String? = null, cpuOnly: Boolean = true): OrtSession {
        val label = modelLabel ?: File(modelPath).name
        val file = File(modelPath)
        if (!file.exists()) {
            throw OrtSessionFactoryError.ModelFileMissing(modelPath)
        }
        val options = makeSessionOptions(modelLabel = label, cpuOnly = cpuOnly)
        return try {
            sharedEnv.createSession(modelPath, options)
        } catch (e: OrtException) {
            throw OrtSessionFactoryError.SessionCreationFailed(label, e)
        }
    }

    /**
     * Creates a session from bytes read out of the app's `assets/` directory (used by the
     * on-device pose backend). Pass `cpuOnly = true` for a CPU-only session (required for the
     * YOLOX detector).
     */
    @Throws(OrtSessionFactoryError::class)
    fun makeSession(
        assetManager: AssetManager,
        assetName: String,
        modelLabel: String? = null,
        cpuOnly: Boolean = true
    ): OrtSession {
        val label = modelLabel ?: assetName
        val bytes = try {
            assetManager.open(assetName).use { it.readBytes() }
        } catch (e: IOException) {
            throw OrtSessionFactoryError.ModelFileMissing("$assetName (in assets)")
        }
        val options = makeSessionOptions(modelLabel = label, cpuOnly = cpuOnly)
        return try {
            sharedEnv.createSession(bytes, options)
        } catch (e: OrtException) {
            throw OrtSessionFactoryError.SessionCreationFailed(label, e)
        }
    }

    /**
     * Builds session options. `cpuOnly = true` leaves the default CPU execution provider only
     * (no accelerated EP appended) — required for YOLOX, whose baked-in NMS has a dynamic-shape
     * node accelerated EPs reject (hard-fails when there are zero detections, common on real
     * footage). `cpuOnly = false` is a placeholder for a future accelerated-EP path (e.g. NNAPI
     * for RTMPose) and currently behaves identically to CPU-only until that is wired.
     */
    private fun makeSessionOptions(modelLabel: String, cpuOnly: Boolean): OrtSession.SessionOptions {
        val options = OrtSession.SessionOptions()
        if (cpuOnly) {
            android.util.Log.i(TAG, "$modelLabel: CPU-only.")
        } else {
            android.util.Log.i(TAG, "$modelLabel: no accelerated EP wired yet; running CPU-only.")
        }
        return options
    }

    private const val TAG = "OrtSessionFactory"
}
